### PR TITLE
NAS-122923 / 23.10 / Refactor Cronjob table

### PR DESCRIPTION
**Pull request by denysbutenko**
TODO:

- [x] Add tests for `ix-table-columns-selector`
- [x] Fix some tests for `cron-list`

Known issue:

- Custom Cell rendering does not apply to hidden columns that present in expandable rows. We must extract rendering logic from `ix-table-body` and make it reusable.

For testing, go to System -> Advanced -> Cron Jobs.

### CODE REVIEW

File: src/app/pages/system/advanced/cron/cron-list/cron-list.component.html [hunk 1]
```diff
@@ -0,0 +1,51 @@
+<ng-template #pageHeader>
+  <ix-page-title-header>
+    <ix-search-input (search)="onListFiltered($event)"></ix-search-input>
+
+    <ix-table-columns-selector [columns]="columns" (changed)="columnsChanged()"></ix-table-columns-selector>
```
Comment (undsoft) [inline src/app/pages/system/advanced/cron/cron-list/cron-list.component.html:5]: If you rename `(changed)` to `(columnsChange)` we may be able to use [two way binding](https://angular.io/guide/two-way-binding) here.

File: src/app/pages/system/advanced/cron/cron-list/cron-list.component.ts [hunk 2]
```diff
@@ -1,167 +1,199 @@
...
+    //   propertyName: 'schedule.dom',
+    // }),
+    // textColumn({
+    //   title: this.translate.instant('Month'),
+    //   propertyName: 'schedule.month',
+    // }),
+    // textColumn({
+    //   title: this.translate.instant('Day of Week'),
+    //   propertyName: 'schedule.dow',
+    // }),
```
Comment (undsoft) [inline src/app/pages/system/advanced/cron/cron-list/cron-list.component.ts:79]: Let's keep to one column/row.

File: src/app/pages/system/advanced/cron/cron-list/cron-list.component.ts [hunk 3]
```diff
@@ -1,167 +1,199 @@
...
+  ngAfterViewInit(): void {
+    this.layoutService.pageHeaderUpdater$.next(this.pageHeader);
+  }
+
+  getCronJobs(): void {
+    this.isLoading = true;
+    this.ws.call('cronjob.query').pipe(
+      map((cronjobs) => {
+        return cronjobs.map((job: CronjobRow) => {
+          const cronSchedule = `${job.schedule.minute} ${job.schedule.hour} ${job.schedule.dom} ${job.schedule.month} ${job.schedule.dow}`;
```
Comment (undsoft) [inline src/app/pages/system/advanced/cron/cron-list/cron-list.component.ts:119]: There is `scheduleToCrontab` that you can use.

**Review by undsoft — CHANGES_REQUESTED** [review]
Looks good so far.

I think we should:
1. Prevent user from unselecting all columns.
2. Do not select all columns by default, and select limited set like we used to have.

File: src/app/modules/ix-table2/components/ix-table-columns-selector/ix-table-columns-selector.component.ts [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+  @Output() changed = new EventEmitter();
+  selectedColumns = new SelectionModel<Column<T, ColumnComponent<T>>>(true, []);
+
+  get isAllChecked(): boolean {
+    return this.selectedColumns.selected.length === this.columns.length;
+  }
+
+  constructor(private cdr: ChangeDetectorRef) {}
+
+  ngOnInit(): void {
```
[the remark this answers is not in the document]
Comment (undsoft) [reply src/app/modules/ix-table2/components/ix-table-columns-selector/ix-table-columns-selector.component.ts:30]: Still valid. Best to make sure this component properly responds to `ngOnChanges`.

File: src/app/modules/ix-table2/components/ix-table-columns-selector/ix-table-columns-selector.component.ts [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+
+@UntilDestroy()
+@Component({
+  selector: 'ix-table-columns-selector',
+  templateUrl: './ix-table-columns-selector.component.html',
+  styleUrls: ['./ix-table-columns-selector.component.scss'],
+  changeDetection: ChangeDetectionStrategy.OnPush,
+})
+export class IxTableColumnsSelectorComponent<T = unknown> implements OnInit {
+  @Input() columns!: Column<T, ColumnComponent<T>>[];
```
Comment (undsoft) [inline src/app/modules/ix-table2/components/ix-table-columns-selector/ix-table-columns-selector.component.ts:17]: Exclamation mark doesn't do much in our codebase.

File: src/app/pages/system/advanced/cron/cron-list/cron-list.component.ts [hunk 6]
```diff
@@ -1,167 +1,182 @@
...
-import { AppState } from 'app/store';
-import { selectTimezone } from 'app/store/system-config/system-config.selectors';
+import { LayoutService } from 'app/services/layout.service';
 
 @UntilDestroy()
 @Component({
-  template: '<ix-entity-table [title]="title" [conf]="this"></ix-entity-table>',
-  providers: [TaskService, UserService],
+  templateUrl: './cron-list.component.html',
+  styleUrls: ['./cron-list.component.scss'],
```
Comment (undsoft) [inline src/app/pages/system/advanced/cron/cron-list/cron-list.component.ts:30]: changeDetection: onPush

**Review by undsoft — APPROVED** [review]
When new cron job is added list is not updated.

Other than that looks good.

File: src/app/modules/ix-table2/components/ix-table-body/cells/ix-cell-schedule/ix-cell-schedule.component.spec.ts [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+    component: IxCellScheduleComponent<TestTableData>,
+    imports: [IxTable2Module],
+  });
+
+  beforeEach(() => {
+    spectator = createComponent({
+      props: {
+        propertyName: 'scheduleField',
+        row: { scheduleField: schedule },
+      } as Partial<IxCellScheduleComponent<TestTableData>>,
```
[the remark this answers is not in the document]
Comment (denysbutenko) [reply src/app/modules/ix-table2/components/ix-table-body/cells/ix-cell-schedule/ix-cell-schedule.component.spec.ts:29]: Other specs are used this way too. `setRow` is being used below in test case.

[the remark this answers is not in the document]
Comment (undsoft) [reply src/app/modules/ix-table2/components/ix-table-body/cells/ix-cell-schedule/ix-cell-schedule.component.spec.ts:29]: Other specs are incorrect too :)

**Pull request merged**